src/main/scala/com/starrocks/connector/spark/read/AbstractStarRocksDecoder.scala

```diff
@@ -0,0 +1,76 @@
+// Modifications Copyright 2021 StarRocks Limited.
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package com.starrocks.connector.spark.read
+
+import com.starrocks.connector.spark.cfg.ConfigurationOptions.STARROCKS_READ_FIELD
+import com.starrocks.connector.spark.cfg.Settings
+import com.starrocks.connector.spark.rdd.{BaseValueReader, RpcValueReader}
+import com.starrocks.connector.spark.rest.RpcPartition
+import com.starrocks.connector.spark.sql.schema.StarRocksSchema
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.connector.read.{InputPartition, PartitionReader}
+
+import scala.collection.JavaConverters.asScalaBufferConverter
+
+abstract class AbstractStarRocksDecoder[Record <: StarRocksGenericRow](partition: InputPartition,
+                                                                       settings: Settings,
+                                                                       schema: StarRocksSchema)
+  extends PartitionReader[Record]
+  with Logging {
+  // the reader obtain data from StarRocks BE
+  lazy val reader: BaseValueReader = {
+    val baseValueReader = new RpcValueReader(partition.asInstanceOf[RpcPartition], settings)
+    baseValueReader.init
+    baseValueReader
+  }
+  var rowOrder: Seq[String] = settings.getProperty(STARROCKS_READ_FIELD).split(",")
+  var currentRow: Array[Any] = Array.fill(rowOrder.size)(null)
+
+  def getNextRecord: AnyRef = {
+    if (!hasNext) {
+      return null
+    }
+    reader.rowBatch.next.asScala.zipWithIndex.foreach {
+      case (s, index) if index < currentRow.size => currentRow.update(index, s)
+      case _ => // nothing
+    }
+    currentRow
+  }
+
+  def hasNext: Boolean = {
+    reader.hasNext
+  }
+
+  def next(): Boolean = {
+    val hasNext = reader.rowBatchHasNext || reader.rowHasNext
```

Comment:
why not use `reader.hasNext` to replace `reader.rowBatchHasNext || reader.rowHasNext`

Reply:
rowBatchHasNext : indicates whether  having the next batch 
rowHasNext : indicates whether  having the next row in current batch

Reply:
I know the meaning of these two variables. I want to say why not encapsulate the  the logic `reader.rowBatchHasNext || reader.rowHasNext` to the function `reader.hasNext`. This similar logic appears many times.

Reply:
resolved

Reply:
reader.hasNext will iterate  rpc request to get next batch. not just as a flag.